deprecate sequence_alignment() and replace deprecated Bio.pairwise2 with Bio.Align.PairwiseAligner

orbeckst opened the pull request:
Fixes #3950 

Changes made in this Pull Request:
- deprecate `sequence_alignment()` and schedule for removal in 3.0
- engineered return tuple as namedtuple to functionally match the originally returned pairwise2.Alignment; we only documented a tuple as return type anyway
- increase biopython required version to >= 1.80
- update CHANGELOG


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

orbeckst committed:
replaced deprecated Bio.pairwise2 with Bio.Align.PairwiseAligner

- fix #3950
- engineered return tuple as namedtuple to functionally match
  the originally returned pairwise2.Alignment; we only documented
  a tuple as return type anyway
- update CHANGELOG

IAlibay commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -1004,22 +1005,46 @@ def sequence_alignment(mobile, reference, match_score=2, mismatch_penalty=-1,
+    # get the old pairwise2 start/stop from the new PairwiseAligner output.
+    return AlignmentTuple(seqA, seqB, topalignment.score,
+                          0, max(reference.n_residues, mobile.n_residues))
 
```
```suggestion


```

IAlibay reviewed (approved):
Assuming the existing tests are sufficient, lgtm! (one missing empty line though :P )

orbeckst commented:
Since biopython 1.80 the output format changed so my hacky way of getting the aligned sequences with gaps is not working anymore, hence the tests fail.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -1004,22 +1005,46 @@ def sequence_alignment(mobile, reference, match_score=2, mismatch_penalty=-1,
+    # extract sequences (there's no obvious way to get the character
+    # representation with gaps by other means from the new
+    # Bio.Align.PairwiseAlignment instance)
+    seqA, _, seqB, _ = topalignment.format().split("\n")
```
This only worked until Bio 1.79. Then they changed the format.

I knew that parsing str was iffy but luckily our tests showed me right away _how_ iffy that was.

I am now trying to find another way to get our old output.

orbeckst replied:
Turns out, it's trivial
```python
seqA = topalignment[0]
seqB = topalignment[1]
```
... testing now.

orbeckst replied:
Needs biopython >= 1.80.

IAlibay replied:
I guess we'll need to offer both options and just do a version check?

IAlibay commented:
> Since biopython 1.80 the output format changed so my hacky way of getting the aligned sequences with gaps is not working anymore, hence the tests fail.

:/ just 1.80? I.e. do we need to support both return types?

orbeckst reviewed (commented):
We do not use `sequence_alignment()` anywhere. Possibly deprecate and remove for 3.0. Might be better as an example in the User Guide anyway. Also, @IAlibay is on a crusade to remove biopython (#3820).

If we ever implement an "automatic" RMSD alignment then we can bring something like it back.

orbeckst committed:
use biopython >= 1.80 feature for updated alig.sequence_alignment()

orbeckst commented:
@IAlibay please review again. 

Sorry for the extra work. When I tested it locally it worked because I had biopython 1.79 locally installed. Only when I updated to 1.80 (as in our CI) it broke. The changes should make it more robust because I _think_ I am using the proper biopython API to get the sequence but it requires the most recent one >= 1.80 (which should be available for Python >= 3.8).

IAlibay commented:
@orbeckst do we have any idea what the minimum dependency versions for Biopython are? I'm apprehensive about having a core dependency that's forced to use the latest release since that can clash with a lot of older deps versions.

IAlibay commented:
The conda-forge feedstock uses pin_compatible numpy so that might be ok? (that's currently set to nep29-ish)

orbeckst committed:
additional docs for sequence_alignment()

(The docs show how one should write code if/when we deprecated this function.)

orbeckst commented:
> do we have any idea what the minimum dependency versions for Biopython are

I tried my code with biopython 1.79 and it failed. For the above to work, it has to be >= 1.80.

orbeckst commented:
Alternatively, we outright deprecate sequence_alignment() and hope that we remove it before biopython removes pairwise2.

orbeckst commented:
Looking at biopython's setup.py https://github.com/biopython/biopython/blob/0d763824ab5ab5fd447fe2589b209e7c64a54b50/setup.py#L112 , they only depend on numpy, any version. Their ci-dependencies.txt is also non-specific.

orbeckst commented:
Not sure where these failures in window come from

[mdanalysis-CI (Azure_Tests Win-Python38-32bit-full)](https://github.com/MDAnalysis/mdanalysis/pull/3951/checks?check_run_id=9956143561) fails [TestH5MDWriterBaseAPI.test_write_trajectory_universe ](https://dev.azure.com/mdanalysis/mdanalysis/_build/results?buildId=4820&view=logs&j=86f9dbba-a936-58d9-7027-9855b63ee954&t=eae8c275-7e36-56e4-0ff0-05635fb44905&l=281)
```
____________ TestH5MDWriterBaseAPI.test_write_trajectory_universe _____________
[gw0] win32 -- Python 3.8.10 C:\hostedtoolcache\windows\Python\3.8.10\x86\python.exe

self = <MDAnalysisTests.coordinates.test_h5md.TestH5MDWriterBaseAPI object at 0x10D5B148>
                            velocities=True, forces=True) as w:
>               for ts in universe.trajectory:

D:\a\1\s\testsuite\MDAnalysisTests\coordinates\test_h5md.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.8.10\x86\lib\site-packages\mdanalysis-2.4.0.dev0-py3.8-win32.egg\MDAnalysis\coordinates\base.py:795: in __iter__
    self._reopen()
C:\hostedtoolcache\windows\Python\3.8.10\x86\lib\site-packages\mdanalysis-2.4.0.dev0-py3.8-win32.egg\MDAnalysis\coordinates\H5MD.py:747: in _reopen
    self.close()
C:\hostedtoolcache\windows\Python\3.8.10\x86\lib\site-packages\mdanalysis-2.4.0.dev0-py3.8-win32.egg\MDAnalysis\coordinates\H5MD.py:728: in close
    self._file.close()
C:\hostedtoolcache\windows\Python\3.8.10\x86\lib\site-packages\h5py\_hl\files.py:435: in close
    file_list = [x for x in file_list if h5i.get_file_id(x).id == self.id.id]
C:\hostedtoolcache\windows\Python\3.8.10\x86\lib\site-packages\h5py\_hl\files.py:435: in <listcomp>
    file_list = [x for x in file_list if h5i.get_file_id(x).id == self.id.id]
h5py\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: Not an id of a file object (not an ID of a file object)

h5py\h5i.pyx:114: ValueError
```

I am letting it *re-run*, maybe that's a transient issue.

IAlibay commented:
> Not sure where these failures in window come from

That's a new one, but I've seen similar transient h5py win32 issues lately. It's been a long while since wheels for win32 got released, honestly we should just kill it soon (definitely for v3.0).

orbeckst commented:
Do you want to deprecate `sequence_alignment()` and remove in 3.0? I can add the deprecation in this PR.

IAlibay commented:
> Do you want to deprecate `sequence_alignment()` and remove in 3.0? I can add the deprecation in this PR.

If it's likely not being used much, then I'd be happy for this to happen.

orbeckst commented:
I am pretty sure I added it a long time ago, in the times of "batteries included is better". It would be fitting if I removed it again ;-). I doubt it's used a lot and in any case, it's 3 lines of code. I'll add deprecations and raise an issue for the UG.

orbeckst committed:
deprecate align.sequence_alignment for removal in 3.0

- add deprecation
- test for deprecation warning
- combined tests for sequence_alignment in a class for easier removal
- update CHANGELOG

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -542,22 +542,38 @@ def test_fasta2select_resids(self, tmpdir):
-        format="string") in seqB, "mobile sequence mismatch"
-    assert_almost_equal(score, 54.6)
-    assert_array_equal([begin, end], [0, reference.n_residues])
+class TestSequenceAlignmentFunction:
```
```suggestion

class TestSequenceAlignmentFunction:
```
Ok I think we have to look into a pep8 action over diffs (easy enough to do on the whole thing, I'm sure it's doable on diffs).

edit: I've gone ahead and opened a PR to do this.

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -542,22 +542,38 @@ def test_fasta2select_resids(self, tmpdir):
+                "`sequence_alignment` will be removed in release 3.0.")
+        with pytest.warns(DeprecationWarning, match=wmsg):
+            align.sequence_alignment(mobile, reference)
 
```
```suggestion


```

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -542,22 +542,38 @@ def test_fasta2select_resids(self, tmpdir):
+                     err_msg="reference sequence mismatch")
+        assert mobile.residues.sequence(
+            format="string") in seqB, "mobile sequence mismatch"
+        assert_almost_equal(score, 54.6)
```
```suggestion
        assert score  == pytest.approx(54.6)
```

IAlibay commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -966,14 +968,16 @@ def rmsd(self):
         warnings.warn(wmsg, DeprecationWarning)
         return self.results.rmsd
 
```
```suggestion


```

IAlibay commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -966,14 +968,16 @@ def rmsd(self):
 
-
+@deprecate(release="2.4.0", remove="3.0",
+           message="See the documentation under Notes how directly use"
```
```suggestion
           message="See the documentation under Notes on how to directly use"
```

IAlibay commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -1002,23 +1006,66 @@ def sequence_alignment(mobile, reference, match_score=2, mismatch_penalty=-1,
-        match_score, mismatch_penalty, gap_penalty, gapextension_penalty)
-    # choose top alignment
-    return aln[0]
+    aligner = Bio.Align.PairwiseAligner(
```
The main argument for pairwisealigner is `scoring` do we want/need to pass that in somehow?

orbeckst replied:
If we are deprecating it then no, let's not add features. At this stage I'd just want to avoid the code suddenly breaking because pairwise2 is gone (and getting rid of DeprecationWarning).

orbeckst commented:
Same error on Windows 32bit. Not sure what to do — no idea which version of hdf5 is installed or if it even has anything to do with bumping biopython to 1.80; at least I wouldn't understand how, given the modest requirements of biopython.

IAlibay commented:
@orbeckst I think I might know where to start looking, do you mind if I push directly to the branch?

orbeckst committed:
Apply suggestions from code review

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

orbeckst commented:
> do you mind if I push directly to the branch?

Please go ahead!

(I can see how reviewing this PR motivated you to get PEP8 back in PR #3954 , sorry.)

IAlibay commented:
Did it... just fix itself?

IAlibay reviewed (approved):
Let's just merge this whilst it's green @orbeckst 🤣 (if azure fails elsewhere we'll do a separate PR)

orbeckst commented:
haha, ok, will do
